exercises/exercise-13/database.ts

```diff
@@ -1,13 +1,85 @@
-export class Database<T> {
+import * as fs from 'fs';
+
+type JsonScalar = boolean | number | string;
+
+export type FieldOp = |
+    {$eq: JsonScalar} |
+    {$gt: JsonScalar} |
+    {$lt: JsonScalar} |
+    {$in: JsonScalar[]};
+
+export type Query<T extends object> = |
+    {$and: Query<T>[]} |
+    {$or: Query<T>[]} |
+    {$text: string} |
+    (
+        {[field in QueryableKeys<T>]?: FieldOp}
+    );
+
+function matchOp(op: FieldOp, v: JsonScalar) {
+    if ('$eq' in op) {
+        return v === op['$eq'];
+    } else if ('$gt' in op) {
+        return v > op['$gt'];
+    } else if ('$lt' in op) {
+        return v < op['$lt'];
+    } else if ('$in' in op) {
+        return op['$in'].includes(v);
+    }
+    throw new Error(`Unrecognized op: ${op}`);
+}
+
+type IndexedRecord<T extends object> = T & {
+    $index: {[word: string]: true};
+};
+
+type Unionize<T extends object> = {[k in keyof T]: {k: k, v: T[k]}}[keyof T];
+type QueryableKeys<T extends object> = Extract<Unionize<T>, {v: JsonScalar}>['k'];
+
+function matches<T extends object>(
+    q: Query<T>,
+    r: IndexedRecord<T>
+): boolean {
+    if ('$and' in q) {
+        return q.$and!.every(subq => matches(subq, r));
+    } else if ('$or' in q) {
+        return q.$or!.some(subq => matches(subq, r));
+    } else if ('$text' in q) {
+        const words = q.$text!.toLowerCase().split(' ');
+        return words.every(w => r.$index[w]);
+    }
+    return Object.entries(q).every(
+        ([k, v]) => matchOp(v as FieldOp, r[k as keyof T] as any)
+    );
+}
+
+export class Database<T extends object> {
     protected filename: string;
-    protected fullTextSearchFieldNames: unknown[];
+    protected fullTextSearchFieldNames: (keyof T)[];
+    protected records: IndexedRecord<T>[];
 
-    constructor(filename: string, fullTextSearchFieldNames) {
+    constructor(filename: string, fullTextSearchFieldNames: (keyof T)[]) {
```

Comment:
It would be nice to enforce that the `fullTextSearchFieldNames` correspond to `string`-valued properties in `T`. One way to do this is with the `Objectify` / `Unionize` generics I defined above (I'll be writing a blog post about these soon):

```ts
type StringKeys<T extends object> = Extract<Unionize<T>, {v: string}>['k'];
export class Database<T extends object> {
  constructor(filename: string, fullTextSearchFieldNames: StringKeys<T>[]) {
    // ...
  }
}
```

It's pretty neat to see this working in autocomplete!

![image](https://user-images.githubusercontent.com/98301/81356590-85f75480-909f-11ea-9d5f-a8c362084298.png)